File: python/LCP_07-2.py
class Solution:
    def numWays(self, n: int, relation: List[List[int]], k: int) -> int:
        dp = [[0] * (k + 1) for _ in range(n)]
        dp[0][0] = 1
        for j in range(1, k + 1):
            for i in range(n):
                for r1, r2 in relation:
                    if r2 == i:
                        dp[r2][j] += dp[r1][j - 1]
        return dp[n - 1][k]

File: LCP_07-2.java
class Solution {
    private boolean[][] edges;
    private int count;
    
    private void visit(int id, int k, int n) {
        if (k == 0) {
            if (id == n - 1) {
                ++count;
            }
            return;
        }
        for (int i = 0; i < n; ++i) {
            if (edges[id][i]) {
                visit(i, k - 1, n);
            }
        }
    }
    
    public int numWays(int n, int[][] relation, int k) {
        edges = new boolean[n][n];
        for (int[] r : relation) {
            edges[r[0]][r[1]] = true;
        }
        count = 0;
        visit(0, k, n);
        return count;
    }
}
